zigpy/util.py

```diff
@@ -135,6 +137,7 @@ def wrapper(*args, tries=tries, delay=delay, **kwargs):
                 retry_exceptions,
                 tries=tries,
                 delay=delay,
+                async_on_failure=async_on_failure,
```

Comment:
`retryable` always wraps methods and if the `async_on_failure` function is also a method, you'd have to add `self` as the first argument of `def wrapper(` to get a reference to it and then "bind" `async_on_failure` to it.

I was trying to implement an API like this:

```Python
    async def on_initialize_failure(self, error=None):
        """
        Called when a device fails to initialize.
        """

        self.application.listener_event("device_init_failure", self)
        await self.application.remove(self.ieee)

    @zigpy.util.retryable([
        asyncio.TimeoutError, zigpy.exceptions.ZigbeeException
    ], tries=3, delay=0.5, failure_callback=on_initialize_failure)
    async def initialize(self):
        ...
```

Reply:
yeah, I forgot that with decorators, passing any instance specific detail is not trivial. Open for ideas.

Reply:
could do it the hacky way, where it is the "callback name" to be called if the args[0] has that member???

Reply:
I'm not seeing a nice way to do this with decorators, tbh. How about using task callbacks?

```Python
        self._initialize_task = asyncio.create_task(self.initialize())

        def done_callback(task):
            try:
                task.result()
            except Exception:
                self.application.listener_event("device_init_failure", self)
                asyncio.create_task(self.application.remove(self.ieee))

        self._initialize_task.add_done_callback(done_callback)

        return self._initialize_task
```

This won't be called if `Device.initialize()` is directly awaited but that seems almost like a feature?

Reply:
So, what about this approach: pass the callback by the name, which would be called if there're args and `args[0]` has a member by that name. 
Is this too hacky?